### PR TITLE
fix(sql): rename log() to ln() natural logarithm function

### PR DESCRIPTION
**Pull request by Ln11211**
Aligned the use of log() function to sql definition of ln according to issue #2578 

-Replaced all log() into ln()
-Renames LogDoubleFunctionFactory class

### CODE REVIEW

**Comment by Ln11211** [comment]
I just changed the `log()` definition into `ln()` according to the sql definition. Couldn't find any `log base 10` definition(and uses)..

**Comment by Ln11211** [comment]
Hey @pswu11 , I searched for all `log()` implementations and changed them to `ln()` so that it aligns with the Postgres definition of `ln()`. This is al that needs to be done? :)

 I was wondering about the `log` and `log10` implementations and I see there is a PR on `log10` that is about to be merged and that won't need any changes whereas a `log(b,a)` ( **logarithm of a to base b** ) implementation is still missing... Can I add that definition?

**Comment by pswu11** [comment]
Hi @Ln11211 thanks a lot, if you want to implement Log(b, a), for sure. 
As for this PR, I'll let my colleagues (core engineers) judge if there's anything else missing.

**Comment by Ln11211** [comment]
> Hi @Ln11211 thanks a lot, if you want to implement Log(b, a), for sure. As for this PR, I'll let my colleagues (core engineers) judge if there's anything else missing.

Thenks !!, I'll implement the log(b,a) through this PR itself, although please confirm with them if it's actually needed or not.. :)

**Comment by Ln11211** [comment]
hello @pswu11 , any updates?

**Comment by pswu11** [comment]
Please go ahead:-)

**Comment by Ln11211** [comment]
Previous definition of `log(num)` in Questdb meant `ln(num)` ,so refactored that and aligned the definition. Now `ln(num)` in Questdb-use means natural logarithm and `log(b,a)` means log a to base b.

**Comment by Ln11211** [comment]
I have to discuss about the implementation of `log(b,a)` which is leading to a rounding off error in a special case.

image
Using the change of base formula(^^^) for certain pair of values that are perfect exponents, such as `log(100,1000000)` the answer is **3** but the division results in **2.9999999999999996** this happens with only certain values of perfect exponents. I have tested several values and I have observed that this accuracy error happens with perfect squares(**less likely**) and perfect cubes(**most likely**)..

By perfect squares and perfect cubes I mean 4 and 8 which are perfect square and cube of 2 etc..

What do we do with this accuracy problem for this special case?

**Comment by Ln11211** [comment]
### All cases except this particular case pass, that too only for cubic.
image
image

### Tried absurdly huge power of 6 and the test passes ###
image

**Comment by Ln11211** [comment]
> I have to discuss about the implementation of `log(b,a)` which is leading to a rounding off error in a special case.
> 
> image Using the change of base formula(^^^) for certain pair of values that are perfect exponents, such as `log(100,1000000)` 

Hello @puzpuzpuz , what can we do with this error?

**Comment by puzpuzpuz** [comment]
That's how Postgres does it - notice the scale manipulations. I'm afraid that we need something similar if we want to support `log(a, b)`.
https://github.com/postgres/postgres/blob/f14aad5169baa5e2ac25d49f1d18f9d5cb3bc7f2/src/backend/utils/adt/numeric.c#L10256

**Comment by puzpuzpuz** [comment]
One more thing to mention - we're not obliged to support everything Postgres has. So, feel free to skip `log(a, b)` function completely. I don't think this variation of the log function is of high demand, anyway.

**Comment by Ln11211** [comment]
> One more thing to mention - we're not obliged to support everything Postgres has. So, feel free to skip `log(a, b)` function completely. I don't think this variation of the log function is of high demand, anyway.

Ooofff :-) . I spent the last night trying to understand what are `scale` and `dweight` are and then checked multiple stackoverflows for log(a,b) . Its interesting though.

**Comment by Ln11211** [comment]
Hello @puzpuzpuz,
  I am also trying the `width_bucket()` implementation and I have come across some major setbacks, firstly a `width_bucket` will take 4 inputs so the sql query will have 4 inputs and we don't have a function to parse (or I guess to convert the query to number, like "D" to "Double" thingy..) , we have at the most a `binary function` and `Double function` implementation for a query of two inputs so I am planning on doing the following-
    **->Create a 4 input sql parser  (sql query 4 input converter to numeric thingy)**
    **->name them as `quaternary function` and ``quadruple function`
 
 is there any other way to get 4 inputs using a sql query without creating these new functions(I have checked binary and unary functions and they are kinda big but easier to implement)

**Comment by puzpuzpuz** [comment]
> is there any other way to get 4 inputs using a sql query without creating these new functions(I have checked binary and unary functions and they are kinda big but easier to implement)

Not sure if I follow the question. Could you explain what you're trying to achieve?

**Comment by Ln11211** [comment]
> > is there any other way to get 4 inputs using a sql query without creating these new functions(I have checked binary and unary functions and they are kinda big but easier to implement)
> 
> Not sure if I follow the question. Could you explain what you're trying to achieve?

It is harder to explain so I attached some screenshots and left comments in the code for you..

This is how double inputs are processed
image

This is how I plan to process 4 inputs
image


My question is
How to take 4 inputs and process them to work on them?, is there a function to do so?

width bucket uses 4 inputs and goes like this `width_bucket(double  number, double least, double highest, int count)` so a sql query will be `width_bucket(20,1,100,2)` ..

**Comment by puzpuzpuz** [comment]
Try using `width_bucker(DDDi)` or even `width_bucker(Dddi)` signature. This way you would be able to read the constant argument values in place and use the `TernaryFunction` (or `UnafryFunction` in case of `Dddi`) interface. Uppercase letters stand for non-constant expressions, such as columns or column expressions or runtime constants (bind variables). You need consider that when implementing a new function.

**Comment by Ln11211** [comment]
> Try using `width_bucker(DDDi)` or even `width_bucker(Dddi)` signature. This way you would be able to read the constant argument values in place and use the `TernaryFunction` (or `UnafryFunction` in case of `Dddi`) interface. Uppercase letters stand for non-constant expressions, such as columns or column expressions or runtime constants (bind variables). You need consider that when implementing a new function.

I'll try them and let you know.

yeah the idea of new function seems a bit difficult.

**Comment by Ln11211** [comment]
> Try using `width_bucker(DDDi)` or even `width_bucker(Dddi)` signature. This way you would be able to read the constant argument values in place and use the `TernaryFunction` (or `UnafryFunction` in case of `Dddi`) interface. Uppercase letters stand for non-constant expressions, such as columns or column expressions or runtime constants (bind variables). You need consider that when implementing a new function.

hello @puzpuzpuz , hope I am not disturbing you, I guess I'll need your help with the signature and taking the input, I had no luck with the 'Ternary' and 'Unary' .

**Comment by puzpuzpuz** [comment]
> hello @puzpuzpuz , hope I am not disturbing you, I guess I'll need your help with the signature and taking the input, I had no luck with the 'Ternary' and 'Unary' .

Is the code available somewhere? You could create a draft PR, so it's easier to follow the code.

**Comment by Ln11211** [comment]
> > hello @puzpuzpuz , hope I am not disturbing you, I guess I'll need your help with the signature and taking the input, I had no luck with the 'Ternary' and 'Unary' .
> 
> Is the code available somewhere? You could create a draft PR, so it's easier to follow the code.

Hey actually someone raised a PR for the `width_bucket` so I guess its solved (the code looks pretty solid ) and surprisingly they implemented the `quaternaryFunction` too. I thank you for you assistance though :)

**Comment by puzpuzpuz** [comment]
I missed that PR.

> surprisingly they implemented the quaternaryFunction too

I don't think we need it in this case, so I'm not convinced.

**Comment by Ln11211** [comment]
> I missed that PR.
> 
> > surprisingly they implemented the quaternaryFunction too
> 
> I don't think we need it in this case, so I'm not convinced.

I'll raise a draft PR then, I am completing the code as I write to you

**Comment by puzpuzpuz** [comment]
> I'll raise a draft PR then, I am completing the code as I write to you

I don't think it's a good idea to raise competing PRs, but it's up to you to decide. We have lots of other good-first-issue tasks available for contributors.

**Comment by Ln11211** [comment]
> > I'll raise a draft PR then, I am completing the code as I write to you
> 
> I don't think it's a good idea to raise competing PRs, but it's up to you to decide. We have lots of other good-first-issue tasks available for contributors.

yup, I agree

**Comment by Ln11211** [comment]
Hey @puzpuzpuz , I was going through the open issues in #2578 , is `floor` yet to be done? I have seen a `floor` already in the code base though.

Also there are two implementations of `ceil` in the codebase with the names `ceil` and `ceiling` that perform the same operation. I believe it's been implemented twice :)

image

**Comment by puzpuzpuz** [comment]
> Hey @puzpuzpuz , I was going through the open issues in #2578 , is `floor` yet to be done? I have seen a `floor` already in the code base though.

Yes, that's true. I don't know why we have it there. @pswu11 shall we mark it as done?

> Also there are two implementations of `ceil` in the codebase with the names `ceil` and `ceiling` that perform the same operation. I believe it's been implemented twice :)

That should have been an alias, not a copy-paste :(

If you'd like to contribute, it might be a good idea to reuse the code in a similar way to the `LowerFunctionFactory` class.

**Comment by Ln11211** [comment]
>If you'd like to contribute, it might be a good idea to reuse the code in a similar way to the LowerFunctionFactory class.

I don't understand what `Lower`  does ?, is it a string function or is it just another name for floor?

**Comment by puzpuzpuz** [comment]
> I don't understand what `Lower` does ?, is it a string function or is it just another name for floor?

It's functionality doesn't matter. I was referring to the fact that it extends `ToLowercaseFunctionFactory` class and overrides `getSignature()` method.

**Comment by Ln11211** [comment]
> > I don't understand what `Lower` does ?, is it a string function or is it just another name for floor?
> 
> It's functionality doesn't matter. I was referring to the fact that it extends `ToLowercaseFunctionFactory` class and overrides `getSignature()` method.

ohhh, yeah yeah , I understand it now, I'll do it for the ceiling, do you want an alias for the floor too?( I guess floor is simple by itself , but "flooring" ? ? for floor kinda sound nice too)

**Comment by puzpuzpuz** [comment]
No, we don't need an alias for `floor`. The only reason to have that weird alias for `ceil` is Postgres compatibility. For some (probably, historical) reasons it has the alias. See:
https://www.postgresql.org/docs/15/functions-math.html

**Comment by Ln11211** [comment]
> No, we don't need an alias for `floor`. The only reason to have that weird alias for `ceil` is Postgres compatibility. For some (probably, historical) reasons it has the alias. See: https://www.postgresql.org/docs/15/functions-math.html

yup makes sense

**Comment by Ln11211** [comment]
@puzpuzpuz yup its done, Now I need your help with something that's been really irking, the "partition issue", the one where a long number is being printed with the table name. I'll ping you there

File: core/src/main/java/io/questdb/griffin/engine/functions/math/CeilingDoubleFunctionFactory.java [hunk 1]
```diff
@@ -34,33 +34,12 @@
 import io.questdb.std.IntList;
 import io.questdb.std.ObjList;
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/CeilingDoubleFunctionFactory.java:35]: Would be nice to remove unused imports here.

File: core/src/main/java/io/questdb/griffin/engine/functions/math/CeilingFloatFunctionFactory.java [hunk 2]
```diff
@@ -33,34 +33,12 @@
 import io.questdb.griffin.engine.functions.UnaryFunction;
 import io.questdb.std.IntList;
 import io.questdb.std.ObjList;
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/CeilingFloatFunctionFactory.java:35]: Would be nice to remove unused imports here.

File: core/src/main/java/io/questdb/griffin/engine/functions/math/CeilingFloatFunctionFactory.java [hunk 3]
```diff
@@ -33,34 +33,12 @@
...
-
-public class CeilingFloatFunctionFactory implements FunctionFactory {
+/**
+ * Postgres-compatibility ceiling(),an alias for the ceil() function.
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/CeilingFloatFunctionFactory.java:37]:
```suggestion
 * Postgres-compatibility ceiling() alias for the ceil() function.
```

File: core/src/main/java/io/questdb/griffin/engine/functions/math/CeilingFloatFunctionFactory.java (hunk 2 above)
Comment (Ln11211) [reply core/src/main/java/io/questdb/griffin/engine/functions/math/CeilingFloatFunctionFactory.java:35]: yeah I overlooked them, gimme a min

**Review by puzpuzpuz — APPROVED** [review]

**Pull request merged**